tests/test_client/tests.py

```diff
@@ -902,6 +903,28 @@ def _generic_view(request):
     return HttpResponse(status=200)
 
 
+urlpatterns = [
+    path('', _generic_view, name='test-client-generic-view')
+]
+
+
+def urlconf_override_middleware(get_response):
+    def middleware(request):
+        request.urlconf = __name__
+        return get_response(request)
+    return middleware
+
+
+@override_settings(ROOT_URLCONF='test_client.urls')
+class URLConfSetByMiddlewareTest(TestCase):
+    def test_resolver_match_when_urlconf_modified_by_middleware(self):
+        middleware = [f'{__name__}.urlconf_override_middleware'] + settings.MIDDLEWARE
+        with override_settings(MIDDLEWARE=middleware):
+            response = self.client.get('/')
+
+        self.assertIs(response.resolver_match.func, _generic_view)
```

Comment:
I would assert against `url_name`, e.g.
```suggestion
        self.assertEqual(response.resolver_match.url_name, 'overridden_urlconf_view')
```